File: android/sample-ui/src/main/kotlin/com/usesmileid/sampleapps/ui/state/UseSmileIDSampleProfiles.kt
package com.usesmileid.sampleapps.ui.state

import androidx.compose.runtime.Immutable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.compose.runtime.toMutableStateList

/** One persona a job runs as: the organisation the SDK's consent screen names, and the details its jobs carry. */
@Immutable
data class UseSmileIDSampleProfile(
    val id: String,
    /** May be blank, when consent names the app itself rather than the person being verified. */
    val organisation: String,
    val defaults: UseSmileIDSampleUserDetails = UseSmileIDSampleUserDetails(),
    /** Empty means the partner's portal default. */
    val callbackUrl: String = "",
) {
    /** The person the details name, so it can never disagree with them. */
    val person: String get() = "${defaults.firstName} ${defaults.lastName}".trim()

    /** What a row calls it: the organisation, or the person when it names none. */
    val title: String get() = organisation.ifBlank { person }.ifBlank { UNNAMED_PROFILE }

    /** The person's initials, as the design has them, falling back to the organisation. */
    val initials: String
        get() = person.ifBlank { organisation }
            .split(" ").filter { it.isNotBlank() }.take(2)
            .joinToString("") { it.first().uppercase() }
            .ifEmpty { "?" }

    /** What a row says under the organisation: the person, or a placeholder until details are saved. */
    val caption: String
        get() = person.ifBlank { NO_USER_DETAILS_CAPTION }
}

/** What is stored: the profiles and which is active. Null [activeId] exactly when there are none. */
@Immutable
data class UseSmileIDSampleProfilesRecord(
    val profiles: List<UseSmileIDSampleProfile> = emptyList(),
    val activeId: String? = profiles.firstOrNull()?.id,
)

/** The profiles the app can act as, and which one is active; a plain first launch has none. [onChange] stores each change, null for a launch that writes nothing. */
class UseSmileIDSampleProfiles(
    seed: UseSmileIDSampleProfilesRecord = UseSmileIDSampleProfilesRecord(),
    loaded: Boolean = true,
    private val onChange: ((UseSmileIDSampleProfilesRecord) -> Unit)? = null,
) {

    private val items = seed.profiles.toMutableStateList()

    var activeId: String? by mutableStateOf(seed.validActiveId())
        private set

    /** False until the stored profiles have been read, so a form does not fill from a list still loading. */
    var loaded by mutableStateOf(loaded)
        private set

    /** The last profile [add] created without activating, until the list that offers "Make active" consumes it. */
    var lastCreatedId: String? by mutableStateOf(null)
        private set

    val all: List<UseSmileIDSampleProfile> get() = items

    val active: UseSmileIDSampleProfile? get() = items.firstOrNull { it.id == activeId }

    /** Position in the list, which is what picks a profile's avatar hue. */
    val activeIndex: Int get() = items.indexOfFirst { it.id == activeId }.coerceAtLeast(0)

    /** What the consent screen names as the partner: the app's own name when no profile names one. */
    val partnerName: String get() = active?.organisation?.takeIf { it.isNotBlank() } ?: NO_PROFILE_PARTNER_NAME

    /** The id a job runs under without a token: the first profile's own id when there is none yet. */
    val partnerId: String get() = active?.id ?: FIRST_PROFILE_ID

    val record: UseSmileIDSampleProfilesRecord get() = UseSmileIDSampleProfilesRecord(items.toList(), activeId)

    /** Adopts what the store read, once; a sign-out tapped before it arrived still wins. */
    fun restore(stored: UseSmileIDSampleProfilesRecord) {
        if (loaded) return
        items.clear()
        activeId = null
        loaded = true
        if (clearedBeforeLoad) {
            changed()
            return
        }
        items.addAll(stored.profiles)
        activeId = stored.validActiveId()
    }

    private var clearedBeforeLoad = false

    fun setActive(id: String) {
        if (items.none { it.id == id } || id == activeId) return
        activeId = id
        changed()
    }

    fun clearLastCreated() {
        lastCreatedId = null
    }

    fun find(id: String) = items.firstOrNull { it.id == id }

    /** The first profile ever made becomes active, so a list with profiles always has one active. */
    fun add(
        organisation: String,
        defaults: UseSmileIDSampleUserDetails = UseSmileIDSampleUserDetails(),
        activate: Boolean = false,
    ): UseSmileIDSampleProfile {
        // First free id, not one derived from the count: duplicate keys crash the list and double a test id.
        val id = generateSequence(items.size + 1) { it + 1 }
            .map { "p-$it" }
            .first { candidate -> items.none { it.id == candidate } }
        val profile = UseSmileIDSampleProfile(id = id, organisation = organisation.trim(), defaults = defaults)
        items.add(profile)
        if (activate || activeId == null) activeId = id else lastCreatedId = id
        changed()
        return profile
    }

    /** A null argument leaves that part alone. */
    fun update(
        id: String,
        organisation: String? = null,
        defaults: UseSmileIDSampleUserDetails? = null,
        callbackUrl: String? = null,
    ) {
        val index = items.indexOfFirst { it.id == id }
        if (index < 0) return
        val current = items[index]
        val updated = current.copy(
            organisation = organisation?.trim() ?: current.organisation,
            defaults = defaults ?: current.defaults,
            callbackUrl = callbackUrl?.trim() ?: current.callbackUrl,
        )
        if (updated == current) return
        items[index] = updated
        changed()
    }

    /** Deleting the active profile hands over to the first one left, so a list with profiles always has one active. */
    fun delete(id: String) {
        if (!items.removeAll { it.id == id }) return
        if (activeId == id) activeId = items.firstOrNull()?.id
        if (lastCreatedId == id) lastCreatedId = null
        changed()
    }

    /** Sign out: every profile goes, which is how a phone is handed to the next person. */
    fun clear() {
        if (!loaded) clearedBeforeLoad = true
        if (items.isEmpty()) return
        items.clear()
        activeId = null
        lastCreatedId = null
        changed()
    }

    /** Nothing is stored before the store has answered, or a partial list would replace what it holds. */
    private fun changed() {
        if (loaded) onChange?.invoke(record)
    }

    companion object {
        /** The fixtures only when `seedProfiles` asks, and never stored: an automation run must not leave people behind. */
        fun forLaunch(
            args: UseSmileIDSampleLaunchArgs,
            onChange: (UseSmileIDSampleProfilesRecord) -> Unit,
        ) = if (args.seedProfiles) {
            UseSmileIDSampleProfiles(UseSmileIDSampleProfilesRecord(fixtures()))
        } else {
            UseSmileIDSampleProfiles(loaded = false, onChange = onChange)
        }

        /** The three the design's sheet shows. Reached only by the `seedProfiles` launch argument — see `spec/launch-args.json`. */
        fun fixtures() = listOf(
            UseSmileIDSampleProfile(
                id = "p-1",
                organisation = "UpTech Finance",
                defaults = UseSmileIDSampleUserDetails(firstName = "Kwame", lastName = "Asante"),
            ),
            UseSmileIDSampleProfile(
                id = "p-2",
                organisation = "Kazi Microlending",
                defaults = UseSmileIDSampleUserDetails(firstName = "Amina", lastName = "Diallo"),
            ),
            UseSmileIDSampleProfile(
                id = "p-3",
                organisation = "PesaLink",
                defaults = UseSmileIDSampleUserDetails(firstName = "Tunde", lastName = "Okafor"),
            ),
        )

        /** The partner the consent screen names when no profile does. */
        const val NO_PROFILE_PARTNER_NAME = "Smile ID"

        /** What a plain launch has always sent as the partner id, so no profile changes nothing on the wire. */
        const val FIRST_PROFILE_ID = "p-1"

        /** What the header, settings card and form say while there is no profile. */
        const val NO_PROFILE_LABEL = "No profile yet"
    }
}

private fun UseSmileIDSampleProfilesRecord.validActiveId(): String? =
    activeId?.takeIf { id -> profiles.any { it.id == id } } ?: profiles.firstOrNull()?.id

private const val NO_USER_DETAILS_CAPTION = "No user details yet"

/** A profile naming neither an organisation nor a person, which only a token binding both names allows. */
private const val UNNAMED_PROFILE = "Unnamed profile"

File: ios/SampleUI/Sources/SampleUI/State/UseSmileIDSampleProfiles.swift
import Foundation

/// One persona a job runs as: the organisation the SDK's consent screen names, and the details its jobs carry.
public struct UseSmileIDSampleProfile: Equatable, Identifiable, Sendable, Codable {
  public let id: String
  /// May be blank, when consent names the app itself rather than the person being verified.
  public var organisation: String
  public var defaults: UseSmileIDSampleUserDetails
  /// Empty means the partner's portal default.
  public var callbackUrl: String

  public init(
    id: String,
    organisation: String,
    defaults: UseSmileIDSampleUserDetails = UseSmileIDSampleUserDetails(),
    callbackUrl: String = ""
  ) {
    self.id = id
    self.organisation = organisation
    self.defaults = defaults
    self.callbackUrl = callbackUrl
  }

  /// The person the details name, so it can never disagree with them.
  public var person: String {
    "\(defaults.firstName) \(defaults.lastName)".trimmingCharacters(in: .whitespaces)
  }

  /// What a row calls it: the organisation, or the person when it names none.
  public var title: String {
    if !organisation.isBlank {
      return organisation
    }
    return person.isBlank ? Self.unnamed : person
  }

  /// The person's initials, as the design has them, falling back to the organisation.
  public var initials: String {
    let source = person.isBlank ? organisation : person
    let letters = source.split(separator: " ").prefix(2).compactMap { word in
      word.first.map { String($0).uppercased() }
    }
    return letters.isEmpty ? "?" : letters.joined()
  }

  /// What a row says under the organisation: the person, or a placeholder until details are saved.
  public var caption: String {
    person.isBlank ? Self.noUserDetailsCaption : person
  }

  static let noUserDetailsCaption = "No user details yet"

  /// A profile naming neither an organisation nor a person, which only a token binding both names allows.
  static let unnamed = "Unnamed profile"

  enum CodingKeys: String, CodingKey {
    case id, organisation, firstName, lastName, email, phone, callbackUrl
  }

  /// Flat, as the other three apps store it; a missing field reads as empty.
  public init(from decoder: Decoder) throws {
    let values = try decoder.container(keyedBy: CodingKeys.self)
    id = try values.decode(String.self, forKey: .id)
    organisation = try values.decodeIfPresent(String.self, forKey: .organisation) ?? ""
    defaults = try UseSmileIDSampleUserDetails(
      firstName: values.decodeIfPresent(String.self, forKey: .firstName) ?? "",
      lastName: values.decodeIfPresent(String.self, forKey: .lastName) ?? "",
      email: values.decodeIfPresent(String.self, forKey: .email) ?? "",
      phone: values.decodeIfPresent(String.self, forKey: .phone) ?? ""
    )
    callbackUrl = try values.decodeIfPresent(String.self, forKey: .callbackUrl) ?? ""
  }

  public func encode(to encoder: Encoder) throws {
    var values = encoder.container(keyedBy: CodingKeys.self)
    try values.encode(id, forKey: .id)
    try values.encode(organisation, forKey: .organisation)
    try values.encode(defaults.firstName, forKey: .firstName)
    try values.encode(defaults.lastName, forKey: .lastName)
    try values.encode(defaults.email, forKey: .email)
    try values.encode(defaults.phone, forKey: .phone)
    try values.encode(callbackUrl, forKey: .callbackUrl)
  }
}

/// The profiles the app can act as, and which is active. A plain first launch has none: no profile is a state of its own, never an empty placeholder that reads as one already set up.
public struct UseSmileIDSampleProfiles: Equatable, Sendable {
  public private(set) var all: [UseSmileIDSampleProfile]
  /// Nil exactly when there are no profiles.
  public private(set) var activeId: String?

  /// The last profile `add` created without activating, until the list that offers "Make active" consumes it.
  public private(set) var lastCreatedId: String?

  public init(_ profiles: [UseSmileIDSampleProfile] = [], activeId: String? = nil) {
    var seen = Set<String>()
    all = profiles.filter { seen.insert($0.id).inserted }
    self.activeId = activeId.flatMap { id in all.contains { $0.id == id } ? id : nil } ?? all.first?.id
  }

  public var active: UseSmileIDSampleProfile? {
    all.first { $0.id == activeId }
  }

  /// Position in the list, which is what picks a profile's avatar hue.
  public var activeIndex: Int {
    all.firstIndex { $0.id == activeId } ?? 0
  }

  /// What the consent screen names as the partner: the app's own name when no profile names one.
  public var partnerName: String {
    guard let organisation = active?.organisation, !organisation.isBlank else { return Self.noProfilePartnerName }
    return organisation
  }

  /// The id a job runs under without a token: the first profile's own id when there is none yet.
  public var partnerId: String {
    active?.id ?? Self.firstProfileId
  }

  public mutating func setActive(_ id: String) {
    if all.contains(where: { $0.id == id }) {
      activeId = id
    }
  }

  public mutating func clearLastCreated() {
    lastCreatedId = nil
  }

  public func find(_ id: String) -> UseSmileIDSampleProfile? {
    all.first { $0.id == id }
  }

  /// The first profile ever made becomes active, so a list with profiles always has one active.
  @discardableResult
  public mutating func add(
    organisation: String,
    defaults: UseSmileIDSampleUserDetails = UseSmileIDSampleUserDetails(),
    activate: Bool = false
  ) -> UseSmileIDSampleProfile {
    // First free id, not one derived from the count: duplicate keys crash the list and double a test id.
    var number = all.count + 1
    while all.contains(where: { $0.id == "p-\(number)" }) {
      number += 1
    }
    let profile = UseSmileIDSampleProfile(
      id: "p-\(number)",
      organisation: organisation.trimmingCharacters(in: .whitespaces),
      defaults: defaults
    )
    all.append(profile)
    if activate || activeId == nil {
      activeId = profile.id
    } else {
      lastCreatedId = profile.id
    }
    return profile
  }

  /// A nil argument leaves that part alone.
  public mutating func update(
    _ id: String,
    organisation: String? = nil,
    defaults: UseSmileIDSampleUserDetails? = nil,
    callbackUrl: String? = nil
  ) {
    guard let index = all.firstIndex(where: { $0.id == id }) else { return }
    if let organisation {
      all[index].organisation = organisation.trimmingCharacters(in: .whitespaces)
    }
    if let defaults {
      all[index].defaults = defaults
    }
    if let callbackUrl {
      all[index].callbackUrl = callbackUrl.trimmingCharacters(in: .whitespacesAndNewlines)
    }
  }

  /// Deleting the active profile hands over to the first one left, so a list with profiles always has one active.
  public mutating func delete(_ id: String) {
    all.removeAll { $0.id == id }
    if activeId == id {
      activeId = all.first?.id
    }
    if lastCreatedId == id {
      lastCreatedId = nil
    }
  }

  /// Sign out: every profile goes, which is how a phone is handed to the next person.
  public mutating func clear() {
    self = UseSmileIDSampleProfiles()
  }

  /// The fixtures only when `seedProfiles` asks; the shell never stores them, so an automation run leaves nobody behind.
  public static func forLaunch(seedProfiles: Bool, stored: UseSmileIDSampleProfiles) -> UseSmileIDSampleProfiles {
    seedProfiles ? UseSmileIDSampleProfiles(fixtures()) : stored
  }

  /// The partner the consent screen names when no profile does.
  public static let noProfilePartnerName = "Smile ID"

  /// What a plain launch has always sent as the partner id, so no profile changes nothing on the wire.
  public static let firstProfileId = "p-1"

  /// What the header, settings card and form say while there is no profile.
  public static let noProfileLabel = "No profile yet"

  /// The three the design's sheet shows. Reached only by the `seedProfiles` launch argument.
  public static func fixtures() -> [UseSmileIDSampleProfile] {
    [
      UseSmileIDSampleProfile(
        id: "p-1",
        organisation: "UpTech Finance",
        defaults: UseSmileIDSampleUserDetails(firstName: "Kwame", lastName: "Asante")
      ),
      UseSmileIDSampleProfile(
        id: "p-2",
        organisation: "Kazi Microlending",
        defaults: UseSmileIDSampleUserDetails(firstName: "Amina", lastName: "Diallo")
      ),
      UseSmileIDSampleProfile(
        id: "p-3",
        organisation: "PesaLink",
        defaults: UseSmileIDSampleUserDetails(firstName: "Tunde", lastName: "Okafor")
      )
    ]
  }
}

public extension UseSmileIDSampleProfiles {
  /// Continue's write-back: the typed details go to the active profile, or become a new active one when there is none. A field the token supplies is never stored, since its value belongs to the token.
  mutating func keep(
    _ details: UseSmileIDSampleUserDetails,
    organisation: String,
    requirement: UseSmileIDSampleUserDetailsRequirement = UseSmileIDSampleUserDetailsRequirement()
  ) {
    let current = active
    let kept = UseSmileIDSampleUserField.allCases.reduce(current?.defaults ?? UseSmileIDSampleUserDetails()) { stored, field in
      requirement.supplies(field) ? stored : field.write(stored, field.read(details))
    }
    if let current {
      update(current.id, defaults: kept)
    } else {
      add(organisation: organisation, defaults: kept, activate: true)
    }
  }
}

/// The stored form of the profiles, one JSON value shared by all four apps so a record reads the same in each.
public enum UseSmileIDSampleProfilesCodec {
  public static let version = 1

  private struct Record: Encodable {
    var version: Int
    var activeId: String?
    var profiles: [UseSmileIDSampleProfile]
  }

  /// Read one profile at a time, so a single bad entry drops that entry rather than every profile.
  private struct StoredRecord: Decodable {
    var version: Int
    var activeId: String?
    var profiles: [Lossy]
  }

  private struct Lossy: Decodable {
    var profile: UseSmileIDSampleProfile?

    init(from decoder: Decoder) throws {
      profile = try? UseSmileIDSampleProfile(from: decoder)
    }
  }

  public static func encode(_ profiles: UseSmileIDSampleProfiles) -> Data? {
    let encoder = JSONEncoder()
    encoder.outputFormatting = .sortedKeys
    return try? encoder.encode(Record(version: version, activeId: profiles.activeId, profiles: profiles.all))
  }

  /// Anything unreadable, including a version this build does not know, is no profiles: never a crash.
  public static func decode(_ data: Data?) -> UseSmileIDSampleProfiles {
    guard let data, let record = try? JSONDecoder().decode(StoredRecord.self, from: data), record.version == version else {
      return UseSmileIDSampleProfiles()
    }
    let profiles = record.profiles.compactMap(\.profile).filter { !$0.id.isBlank }
    return UseSmileIDSampleProfiles(profiles, activeId: record.activeId)
  }
}

/// The new-profile sheet's five fields, held outside the sheet so a tab switch cannot lose them.
public struct UseSmileIDSampleNewProfile: Equatable, Sendable {
  public var name: String
  public var firstName: String
  public var lastName: String
  public var email: String
  public var phone: String

  public init(name: String = "", firstName: String = "", lastName: String = "", email: String = "", phone: String = "") {
    self.name = name
    self.firstName = firstName
    self.lastName = lastName
    self.email = email
    self.phone = phone
  }

  /// The design's own rule: Create needs the profile name and both required names.
  public var canCreate: Bool {
    !name.isBlank && !firstName.isBlank && !lastName.isBlank
  }

  /// The person is the two required names; all four seed the details its jobs start from.
  public var person: String {
    "\(firstName) \(lastName)".trimmingCharacters(in: .whitespaces)
  }

  public var defaults: UseSmileIDSampleUserDetails {
    UseSmileIDSampleUserDetails(firstName: firstName, lastName: lastName, email: email, phone: phone)
  }
}
